protocol/client/messenger_v2.go

```diff
@@ -158,18 +163,46 @@ func (m *MessengerV2) Request(ctx context.Context, c Contact, options protocol.R
 }
 
 func (m *MessengerV2) RequestAll(ctx context.Context, newest bool) error {
-	contacts, err := m.db.Contacts()
+	// FIXME(dshulyak) if newest is false request 24h?
+	histories, err := m.db.Histories()
 	if err != nil {
 		return errors.Wrap(err, "error fetching contacts")
 	}
-	requestParams := protocol.DefaultRequestOptions()
-	for _, c := range contacts {
-		err = enhanceRequestOptions(c, &requestParams)
+	now := time.Now()
+	sync, nosync := splitIntoSyncedNotSynced(histories)
+	errors := make(chan error, 2)
+	waitRequest := func(histories []History, opts protocol.RequestOptions) {
+		log.Printf("[messenger::RequestAll] requesting messages for chats %+v: from %d to %d\n", opts.Chats, opts.From, opts.To)
+		start := time.Now()
+		err := m.proto.Request(ctx, opts)
+		if err != nil {
+			errors <- err
+		}
+		log.Printf("[messenger::RequestAll] requesting message for chats %+v finished. took %v\n", opts.Chats, time.Since(start))
+		for i := range histories {
+			histories[i].Synced = now.Unix()
+		}
+		err = m.db.UpdateHistories(histories)
+		errors <- err
+	}
+	wait := 0
```

Comment:
`sync.WaitGroup`?

I mean:
```go
if len(histories) == 0 {
  return 
}

var wg sync.WaitGroup

if len(sync) != 0 {
  wg.Add(1)
  go waitRequest(sync, syncedToOpts(sync, now))
}

// the same for nosync

wg.Wait()

select {
case err := <-errors:
  return err
default:
  return nil
}
```

I think it's beneficial to wait for all goroutines to finish here. Otherwise, we can have a leak and never know because this code would never hang.